Adda a helm chart for in-cluster kubernetes_service agent

awly opened the pull request:
This is a simplified version of the teleport chart, intended to only run
a "stateless" `kubernetes_service` instance within a kubernetes cluster.
This instance joins an externally-managed teleport cluster, given a
proxy address and a join token. The connection is always over a reverse
tunnel, per our recommended approach.

The chart is opinionated and only lets the user modify the bare minimum.

Usage:
```
$ helm install teleport-agent . \
  --namespace teleport \
  --set proxyAddr=$PROXY_ENDPOINT \
  --set authToken=$JOIN_TOKEN \
  --set kubeClusterName=$KUBERNETES_CLUSTER_NAME
```

Also, add `pods/get` permission to `ClusterRole`s everywhere. It's not mandatory, but lets us collect more info for session audit logs in 5.0

Updates https://github.com/gravitational/teleport/issues/4883

Commit:
Add helm chart for in-cluster kubernetes_service agent

This is a simplified version of the teleport chart, intended to only run
a "stateless" `kubernetes_service` instance within a kubernetes cluster.
This instance joins an externally-managed teleport cluster, given a
proxy address and a join token. The connection is always over a reverse
tunnel, per our recommended approach.

The chart is opinionated and only lets the user modify the bare minimum.

benarent commented:
I was going to ask about why we need cluster name but noticed this. 

> The name of the k8s cluster is specified in the `kubernetes_service.kube_cluster_name` field of teleport.yaml. There is no portable way to guess the k8s cluster name from the environment.

benarent commented on examples/chart/teleport-agent/Chart.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: v2
+version: 0.0.1
+description: Teleport provides a secure SSH and Kubernetes remote access solution that doesn't get in the way.
+icon: https://gravitational.com/gravitational/images/logos/company-logos/teleport-symbol-400x400.png
```
We should make this the new logo, I'll try to get it fro the UX crew.

awly replied:
thanks, changed

benarent commented on examples/chart/teleport-agent/README.md:
```diff
@@ -0,0 +1,35 @@
+- a [static join
+  token](https://goteleport.com/teleport/docs/admin-guide/#adding-nodes-to-the-cluster)
+  for this Teleport cluster (`$JOIN_TOKEN`)
+  - this chart does not currently support dynamic join tokens; please [file an
```
What's the reason for it not supporting dynamic join tokens?

webvictim replied:
I think it should work as long as the user pre-generates the token with `tctl` and passes it in here...

webvictim replied:
Ah, it doesn't work because there's no persistent storage so Teleport will have to rejoin and regenerate certs every time the pod restarts.

A user could work around this by exporting a dynamic join token as a k8s secret and having the chart mount this secret as a file inside the pod which Teleport reads on startup. Whenever the secret is updated, the file will change and the new secret will be used on restart. We could add instructions to the README on how to do this.

Theoretically, though, a static join token with sufficient entropy which is only valid for the `Kubernetes` role should be secure enough.

awly replied:
Yeah, it's all about long-term issues due to pod restarts/reschedules.
Static token is always worse (even with large entropy), but we currently don't have a good mechanism to distribute dynamic tokens within k8s. I want to keep the UX as simple as possible (fire-and-forget 1-liner).

We should think about some tighter integration between teleport and k8s, for example using the k8s API as a backend for some essential teleport state (e.g. post-bootstrap credentials, process UUID, etc).

r0mant commented on examples/chart/teleport-agent/Chart.yaml:
```diff
@@ -0,0 +1,7 @@
+name: teleport-agent
+apiVersion: v2
+version: 0.0.1
```
Consider also adding an `appVersion` field to the manifest to specify Teleport version itself? This way a user will be able to easily see Teleport version when running `helm ls` for example. The downside I guess is that it will have to be updated whenever we bump the image this chart uses.

webvictim replied:
We've recently started using the generic 4.3/4.4/5.0 tagged images for the charts (which are automatically rebuilt every night with the latest version of Teleport in that branch) so this isn't a huge problem. I like the idea in general and might adopt it for the other charts.

awly replied:
Added `appVersion` here and also used it as the image tag in the deployment.
There's still a `teleportVersionOverride` in `values.yaml` if it's needed.

I hope this won't become a maintenance burden - yet another field to update when we make a release.

webvictim replied:
It will, but we can script it.

r0mant commented on examples/chart/teleport-agent/Chart.yaml:
```diff
@@ -0,0 +1,7 @@
+name: teleport-agent
```
Is this chart supposed to be Kubernetes service specific (like readme suggests), or can it potentially be extended to support other services as well? Just wondering whether the name is a bit too generic or not.

awly replied:
Good point, renamed to `teleport-kube-agent`

r0mant commented on examples/chart/teleport-agent/values.yaml:
```diff
@@ -0,0 +1,36 @@
+##################################################
+
+# Join token for the cluster.
+authToken:
```
Any reason you decided to use flat values structure, instead of organizing them in individual sections mirroring Teleport's config, e.g. auth, proxy, kube, etc. Would probably be easier to extend later with other services (app, db).

awly replied:
This makes the CLI usage slightly more awkward:
```
$ helm install teleport-agent . \
  --namespace teleport \
  --set teleport.authServers=$PROXY_ENDPOINT \
  --set teleport.authToken=$JOIN_TOKEN \
  --set kubernetes_service.kubeClusterName=$KUBERNETES_CLUSTER_NAME
```
vs
```
$ helm install teleport-agent . \
  --namespace teleport \
  --set proxyAddr=$PROXY_ENDPOINT \
  --set authToken=$JOIN_TOKEN \
  --set kubeClusterName=$KUBERNETES_CLUSTER_NAME
```

My goal here is to make this chart single-purpose (register a k8s cluster in teleport) and as simple as possible.
We have https://github.com/gravitational/teleport/tree/master/examples/chart/teleport which is more suited to be the extensible universal chart.

webvictim replied:
I think as the chart has one lone purpose it's fine to leave it as-is.

r0mant replied:
Sounds good.

r0mant commented on examples/chart/teleport-agent/values.yaml:
```diff
@@ -0,0 +1,36 @@
+# Number of replicas for the agent deployment.
+replicaCount: 1
+# Name of the ClusterRole, used by the agent's service account.
+clusterRoleName: teleport-agent-cr
```
I'd consider getting rid of redundant suffixes `-cr` here and `-crb` below.

awly replied:
done

r0mant commented on examples/chart/teleport-agent/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+        configMap:
+          name: {{ .Release.Name }}
+      - name: "data"
+        emptyDir: {}
```
Just thinking out loud, nodes (and all our new `*_service`'s) are pretty much stateless, right? So it should be fine to use emptyDir.

webvictim replied:
Our other charts also do this.

awly replied:
they are stateless, except for their credentials (which can be re-generated on restart using a join token)

r0mant commented on examples/chart/teleport-agent/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+        app: {{ .Release.Name }}
+    spec:
+      containers:
+      - name: "teleport"
```
Are there healthz/readyz endpoints on Teleport to point liveness/readiness probes to?

awly replied:
Added `/healthz` as liveness probe and `/readyz` as readiness probe
See https://goteleport.com/teleport/docs/admin-guide/#troubleshooting

r0mant commented on examples/chart/teleport-agent/templates/config.yaml:
```diff
@@ -0,0 +1,20 @@
+data:
+  teleport.yaml: |
+    teleport:
+      auth_token: {{ required "authToken is required in chart values" .Values.authToken }}
```
I would consider creating a Secret with the token instead, mounting it into the container and taking it from there, I think Teleport supports that: https://goteleport.com/teleport/docs/admin-guide/#static-tokens.

awly replied:
This complicates the setup (one extra object, layer of indirection), so I'd like to skip that.
AFAIK, secrets don't provide any real security benefit, except for hinting to humans that it's a sensitive resource.
My knowledge may be out-of-date, so please correct me if it's wrong :)

webvictim replied:
I think this is fine for now and in a future version we can head down the route of dynamic tokens and some other form of provisioning. People who desperately want additional security at the cost of added complexity now can always fork the chart and come up with their own distribution method.

r0mant replied:
> AFAIK, secrets don't provide any real security benefit, except for hinting to humans that it's a sensitive resource.

This is true for default configurations but a couple of things to consider are:

* Cluster may be configured with a KMS provider in which case secrets data will be encrypted at rest.
* Placing sensitive data in secrets lets users to configure granular RBAC policies for example allowing somebody to view configuration but not the secrets.

It may not be the end of the world as it is (esp. if we're gonna change this eventually to use PKI), I guess I personally just don't see a compelling argument _against_ placing it into a secret. An additional object shouldn't be any trouble as it's all managed by Helm.

Feel free to leave as is if you still feel it's not worth it.

awly replied:
RBAC and encryption at rest for Secrets are very good reasons, thanks for pointing those out!
Updated to load join token from a Secret.

benarent commented on examples/chart/teleport-agent/Chart.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: v2
+version: 0.0.1
+description: Teleport provides a secure SSH and Kubernetes remote access solution that doesn't get in the way.
+icon: https://gravitational.com/gravitational/images/logos/company-logos/teleport-symbol-400x400.png
```
```suggestion
icon: https://goteleport.com/images/logos/logo-teleport-square.svg
```

webvictim commented on examples/chart/teleport-kube-agent/Chart.yaml:
```diff
@@ -0,0 +1,8 @@
+name: teleport-kube-agent
+apiVersion: v2
+version: 0.0.1
+appVersion: 5.0.0-rc.2
```
This should probably be 5.0.0, as it will be out by the time people are wanting to use this chart.

webvictim commented on examples/chart/teleport-daemonset/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
-version: 0.0.6
+version: 0.0.7
 description: Teleport daemonset installation which provides some access to underlying Kubernetes nodes.
 icon: https://gravitational.com/gravitational/images/logos/company-logos/teleport-symbol-400x400.png
```
```suggestion
icon: https://goteleport.com/images/logos/logo-teleport-square.svg
```

webvictim commented on examples/chart/teleport-auto-trustedcluster/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
-version: 0.0.6
+version: 0.0.7
 description: Teleport trusted cluster installation which automatically joins itself back to the provided root cluster.
 icon: https://gravitational.com/gravitational/images/logos/company-logos/teleport-symbol-400x400.png
```
```suggestion
icon: https://goteleport.com/images/logos/logo-teleport-square.svg
```

webvictim commented on examples/chart/teleport/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
-version: 0.0.6
+version: 0.0.7
 description: Teleport provides a secure SSH and Kubernetes remote access solution that doesn't get in the way.
 icon: https://gravitational.com/gravitational/images/logos/company-logos/teleport-symbol-400x400.png
```
```suggestion
icon: https://goteleport.com/images/logos/logo-teleport-square.svg
```

webvictim reviewed (approved):
LGTM - just added a few suggestions to update the logo for the other charts at the same time.

Commit:
Apply suggestions from code review

Co-authored-by: Gus Luxton <gus@gravitational.com>

Commit:
Move join token into a secret

Secret can be more tightly restricted via RBAC, and encrypted at rest
with KMSs.

Also, a few other small tweaks for UX.

webvictim commented:
@awly It looks like this PR is missing a backport to `branch/5.0` which is confusing some customers. Could we get it backported?
